client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -102,53 +107,94 @@
     final List<Header> defaultHeaders;
     private final long maxRetryTimeoutMillis;
     private final String pathPrefix;
-    private final AtomicInteger lastHostIndex = new AtomicInteger(0);
-    private volatile HostTuple<Set<HttpHost>> hostTuple;
+    private final AtomicInteger lastNodeIndex = new AtomicInteger(0);
     private final ConcurrentMap<HttpHost, DeadHostState> blacklist = new ConcurrentHashMap<>();
     private final FailureListener failureListener;
+    private volatile NodeTuple<List<Node>> nodeTuple;
 
     RestClient(CloseableHttpAsyncClient client, long maxRetryTimeoutMillis, Header[] defaultHeaders,
-               HttpHost[] hosts, String pathPrefix, FailureListener failureListener) {
+               Node[] nodes, String pathPrefix, FailureListener failureListener) {
         this.client = client;
         this.maxRetryTimeoutMillis = maxRetryTimeoutMillis;
         this.defaultHeaders = Collections.unmodifiableList(Arrays.asList(defaultHeaders));
         this.failureListener = failureListener;
         this.pathPrefix = pathPrefix;
-        setHosts(hosts);
+        setNodes(nodes);
     }
 
     /**
      * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
      * Creates a new builder instance and sets the hosts that the client will send requests to.
+     * <p>
+     * Prefer this to {@link #builder(HttpHost...)} if you have metadata up front about the nodes.
+     * If you don't either one is fine.
+     */
+    public static RestClientBuilder builder(Node... nodes) {
+        return new RestClientBuilder(nodes);
+    }
+
+    /**
+     * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
+     * Creates a new builder instance and sets the nodes that the client will send requests to.
+     * <p>
+     * You can use this if you do not have metadata up front about the nodes. If you do, prefer
+     * {@link #builder(Node...)}.
+     * @see Node#Node(HttpHost)
      */
     public static RestClientBuilder builder(HttpHost... hosts) {
-        return new RestClientBuilder(hosts);
+        return new RestClientBuilder(hostsToNodes(hosts));
     }
 
     /**
-     * Replaces the hosts that the client communicates with.
-     * @see HttpHost
+     * Replaces the nodes with which the client communicates.
      */
-    public synchronized void setHosts(HttpHost... hosts) {
-        if (hosts == null || hosts.length == 0) {
-            throw new IllegalArgumentException("hosts must not be null nor empty");
+    public void setNodes(Node... nodes) {
+        if (nodes == null) {
+            throw new IllegalArgumentException("nodes must not be null or empty");
+        }
+        setNodes(Arrays.asList(nodes));
+    }
+
+    /**
+     * Replaces the nodes with which the client communicates.
+     */
+    public synchronized void setNodes(Collection<Node> nodes) {
```

Comment:
could we settle on only one of these two setNodes methods maybe? Seems like the most used one is this one. I would expect order to matter here (I remember recently making a change to make it matter at least) while accepting a Collection could be perceived as the order does not matter? Though we respect the iteration order I think so we should be ok. Again, thinking out loud. I think it's fine if we keep this method and get rid of the one accept a varargs, I do not have a preference between one or the other.